platform/asterix/boot/src/drivers/pmic.c

```diff
@@ -91,94 +141,83 @@ int pmic_init(void) {
-  // 1.8V -- system FW can reenable hysteretic mode later to save power.
-  ret = prv_pmic_write(BUCK_BASE, BUCK1PWMSET, BUCK1PWMSET_SET);
-  if (ret != 0) {
-    return ret;
-  }
-
-  ret = prv_pmic_write(BUCK_BASE, BUCK1NORMVOUT, BUCKVOUT_1V8);
-  if (ret != 0) {
-    return ret;
-  }
-
-  ret = prv_pmic_write(BUCK_BASE, BUCK1RETVOUT, BUCKVOUT_1V8);
-  if (ret != 0) {
-    return ret;
-  }
-
-  ret = prv_pmic_write(BUCK_BASE, BUCKSWCTRLSET, BUCKSWCTRLSET_BUCK1SWCTRLSET);
-  if (ret != 0) {
-    return ret;
-  }
-
-  // Configure charger (TODO: values are board/battery dependent)
-  // - Thermistor: 10K NTC
-  // - Termination voltage: 4.2V
-  // - Reduced termination voltage (for warm region): 4.00V
-  // - 64mA charge/discharge current (standard charging)
-  // - Enable charging
-  ret = prv_pmic_write(CHARGER_BASE, BCHGENABLECLR, ENABLECHARGING_DISABLECHG);
-  if (ret != 0) {
-    return ret;
-  }
-
-  ret = prv_pmic_write(ADC_BASE, ADCNTCRSEL, ADCNTCRSEL_10K);
-  if (ret != 0) {
-    return ret;
-  }
-
-  ret = prv_pmic_write(CHARGER_BASE, BCHGVTERM, BCHGVTERMNORM_4V20);
-  if (ret != 0) {
-    return ret;
-  }
-
-  ret = prv_pmic_write(CHARGER_BASE, BCHGVTERMR, BCHGVTERMREDUCED_4V00);
-  if (ret != 0) {
-    return ret;
-  }
-
-  ret = prv_pmic_write(CHARGER_BASE, BCHGISETMSB, 16);
-  if (ret != 0) {
-    return ret;
-  }
-
-  ret = prv_pmic_write(CHARGER_BASE, BCHGISETDISCHARGEMSB, 16);
-  if (ret != 0) {
-    return ret;
-  }
-
-  ret = prv_pmic_write(CHARGER_BASE, BCHGENABLESET, ENABLECHARGING_ENABLECHG);
-  if (ret != 0) {
-    return ret;
-  }
-
-  // LDO2 as LDO @ 1.8V (powers the QSPI flash)
-  ret = prv_pmic_write(LDSW_BASE, LDSW2LDOSEL, LDSW2LDOSEL_LDO);
-  if (ret != 0) {
-    return ret;
-  }
-
-  ret = prv_pmic_write(LDSW_BASE, LDSW2VOUTSEL, LDSW2VOUTSEL_1V8);
-  if (ret != 0) {
-    return ret;
-  }
-
-  ret = prv_pmic_write(LDSW_BASE, TASKLDSW2SET, 0x01U);
-  if (ret != 0) {
-    return ret;
+  const struct pmic_reg regs[] = {
+    // Turn off any watchdog / boot timer right away.
+    { TIMER_BASE, TIMERCLR, TIMERCLR_TASKTIMERDIS },
+    { ERRLOG_BASE, SCRATCH0, 0x00 /* contains boot timer bit */ },
+    
+    // Make sure right away that we can reset the device if needed.
+    { SHIP_BASE, LPRESETCFG, 0 },
+    
+    // Set up the BUCK1 regulator for manual control to 1.8V, automatic
+    // PWM/hysteresis control.
+    { BUCK_BASE, BUCK1ENASET, 1 },
+    { BUCK_BASE, BUCK2ENASET, 1 },
+    { BUCK_BASE, BUCK1PWMCLR, 1 },
+    { BUCK_BASE, BUCK2PWMCLR, 1 },
+    { BUCK_BASE, BUCK1NORMVOUT, BUCKVOUT_1V8 },
+    { BUCK_BASE, BUCK1RETVOUT, BUCKVOUT_1V8 },
+    { BUCK_BASE, BUCK2NORMVOUT, BUCKVOUT_3V0 },
+    { BUCK_BASE, BUCK2RETVOUT, BUCKVOUT_3V0 },
+    { BUCK_BASE, BUCKENCTRL, 0 },
+    { BUCK_BASE, BUCKVRETCTRL, 0 },
+    { BUCK_BASE, BUCKPWMCTRL, 0 },
+    { BUCK_BASE, BUCKSWCTRLSET, BUCKSWCTRLSET_BUCK1SWCTRLSET | BUCKSWCTRLSET_BUCK2SWCTRLSET /* use registers rather than resistor settings */ },
```

Comment:
note: vset resistors will always force nominal voltages (1.8 and 3.0V), so it may be unsafe to allow sw controlled values here.

Also, as a general comment: holding the back button for 10s will reset the PMIC, so we have a known path to go back to a safe state in case things go terribly wrong.

Reply:
Well, it will reset the PMIC as long as LPRESETCFG is set properly!

But I would like to explicitly set these values, because 1) analog straps always kind of scare me, and 2) when we set them explicitly, the behavior changes in at least some way (!) vs. when we leave them implicitly set from Rset.